controllers/vrg_volrep.go

```diff
@@ -930,6 +926,82 @@ func (v *VRGInstance) reconcileMissingVR(pvc *corev1.PersistentVolumeClaim, log
 	return vrMissing, !requeue
 }
 
+func (v *VRGInstance) getVolumeGroupReplicationState(state volrep.ReplicationState) (volgroup.ReplicationState, error) {
+	var volGroupRepState volgroup.ReplicationState
+
+	switch state {
+	case volrep.Primary:
+		volGroupRepState = volgroup.Primary
+	case volrep.Secondary:
+		fallthrough
+	case volrep.Resync:
+		volGroupRepState = volgroup.Secondary
+	default:
+		v.log.Info(fmt.Sprintf("invalid Replication State %s", string(state)))
+
+		return volGroupRepState, fmt.Errorf("invalid Replication State %s", string(state))
+	}
+
+	return volGroupRepState, nil
+}
+
+func (v *VRGInstance) getVolumeReplication(pvc *corev1.PersistentVolumeClaim,
+	vrNamespacedName *types.NamespacedName,
+) client.Object {
+	volRep := &volrep.VolumeReplication{}
+
+	volumeReplicationClass, vrTypeErr := v.selectVolumeReplicationClass(pvc)
+	if vrTypeErr != nil {
+		return volRep
+	}
+
+	switch volumeReplicationClass.(type) {
+	case *volgroup.VolumeGroupReplicationClass:
+		vrNamespacedName.Name = volumeReplicationClass.GetName() + "-vgr"
+
+		return &volgroup.VolumeGroupReplication{}
+	default:
+		return volRep
+	}
+}
+
+func (v *VRGInstance) createVolumeReplication(pvc *corev1.PersistentVolumeClaim,
+	state volrep.ReplicationState, volumeReplicationClass client.Object,
+) client.Object {
+	return &volrep.VolumeReplication{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      pvc.Name,
+			Namespace: pvc.Namespace,
+		},
+		Spec: volrep.VolumeReplicationSpec{
+			DataSource: corev1.TypedLocalObjectReference{
+				Kind:     "PersistentVolumeClaim",
+				Name:     pvc.Name,
+				APIGroup: new(string),
+			},
+			ReplicationState:       state,
+			VolumeReplicationClass: volumeReplicationClass.GetName(),
+			AutoResync:             v.autoResync(state),
+		},
+	}
+}
+
+func (v *VRGInstance) createVolumeGroupReplication(pvc *corev1.PersistentVolumeClaim,
+	state volgroup.ReplicationState, volumeReplicationClass client.Object,
+) client.Object {
+	return &volgroup.VolumeGroupReplication{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      volumeReplicationClass.GetName() + "-vgr",
+			Namespace: pvc.Namespace,
+		},
+		Spec: volgroup.VolumeGroupReplicationSpec{
+			ReplicationState:            state,
+			VolumeGroupReplicationClass: volumeReplicationClass.GetName(),
+			Selector:                    &v.instance.Spec.PVCSelector,
```

Comment:
In RBD consistency group, the same pvc selector is used in VolumeGroupReplication with the VolumeReplicationGroup, which means all the pvcs in a VRG should be in a single consistency group?